![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### 👨‍🏫👩‍🏫 Cuerpo Docente:

- Profesor: Sebastián Tinoco, Stefano Schiappacasse
- Auxiliar: Melanie Peña Torres, Valentina Rojas Osorio
- Ayudante: Valentina Zuñiga, Ángelo Muñoz 

### 👨‍💻👩‍💻 Estudiantes:
- Estudiante n°1: Camila Salas R.
- Estudiante n°2: Ignacio Negrete S.

_Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir._

---

---

## 📖 Enunciado 

<div align="center">
    <img src='https://github.com/MDS7202/MDS7202/blob/main/recursos/2025-01/proyecto/proyecto.png?raw=true' style="border-radius: 12px"> 
</div>

En el competitivo universo de las bebidas gaseosas, la empresa **SodAI Drinks 🥤** ha logrado destacarse por su creatividad, diversidad de productos y enfoque centrado en el cliente. Ofrece una extensa gama de bebidas carbonatadas que abarca distintos segmentos del mercado: desde productos premium en presentaciones sofisticadas, hasta gaseosas accesibles para el consumo masivo, disponibles en diversos tamaños y tipos de envases. 

La compañía opera en múltiples regiones y zonas, sirviendo a una variedad de puntos de venta que incluyen desde tiendas de conveniencia y minimarkets hasta el canal frío tradicional. Cada tipo de cliente tiene sus particularidades: algunos reciben entregas hasta 4 veces por semana, mientras que otros son visitados por la fuerza de ventas solo una vez semanalmente. Esta diversidad de perfiles representa tanto una oportunidad como un desafío comercial: ¿cómo saber qué productos tienen más chances de ser comprados por cada cliente en un momento dado?

Con el objetivo de aumentar la facturación de forma inteligente y mejorar la eficiencia de su estrategia de ventas, **SodAI Drinks** decide crear una nueva célula interna de innovación: el equipo **Deep Drinkers 🤖**, cuyo propósito es aplicar ciencia de datos para anticiparse a las necesidades del cliente y potenciar el negocio desde una perspectiva basada en información.

El corazón de esta iniciativa es el desarrollo de un sistema predictivo personalizado para cada cliente. Para ello, **Deep Drinkers** convoca a un equipo de Data Scientists y especialistas en *machine learning* con una misión clara: construir un modelo predictivo que, cada semana, pueda estimar la probabilidad de compra de cada producto del portafolio para cada cliente activo.

El modelo deberá tener en cuenta múltiples factores, incluyendo:
- **Tipo de cliente**, ej. "TIENDA DE CONVENIENCIA", "MINIMARKET".
- **Frecuencia de entregas y visitas**, indicadores del nivel de actividad comercial.
- **Ubicación geográfica** (por región y zona).
- **Preferencias históricas de consumo**, inferidas por patrones de compra anteriores.
- **Características del producto**, como marca, categoría, segmento, tipo de envase y tamaño

El objetivo final es que, **cada semana**, se genere una tabla de productos priorizados: para cada cliente, un listado de productos ordenado por su probabilidad estimada de compra. Esta información será enviada al equipo comercial, que podrá usarla en call center, para incrementar las chances de concretar ventas al ofrecer justo lo que el cliente probablemente quiere comprar.

Este proyecto representa un cambio de paradigma en la forma en que **SodAI Drinks** gestiona su fuerza de ventas: de un enfoque reactivo y generalista, a uno proactivo, basado en datos y profundamente personalizado. Así, la empresa no solo espera aumentar su rentabilidad, sino también construir relaciones más sólidas con sus clientes, ofreciéndoles recomendaciones más relevantes y oportunas.

Para lograr lo anterior, el equipo **Deep Drinkers** contará con los siguientes conjuntos de datos, junto a sus respectivos atributos:

- **Datos transaccionales** (`transacciones.parquet`): contiene el historial de compras realizadas por los clientes.
	- `customer_id`: identificador único del cliente que realizó la compra.
	- `product_id`: identificador único del producto comprado.
	- `purchase_date`: fecha en que se realizó la transacción.
	- `order_id`: identificar de la orden de su pedido.
	- `payment`	monto total pagado por la transacción.

- **Datos de clientes** (`clientes.parquet`): incluye las características de cada cliente.
	- `customer_id`: identificador único del cliente.
	- `region_id`: identificador de la región geográfica donde se encuentra el cliente.
	- `customer_type`: tipo de cliente según el canal comercial, por ejemplo, “TIENDA DE CONVENIENCIA”.
	- `Y`: coordenada geográfica de latitud.
	- `X`: coordenada geográfica de longitud.
	- `num_deliver_per_week`: cantidad de entregas semanales que recibe el cliente.
	- `num_visit_per_week`: frecuencia de visitas de la fuerza de ventas por semana.

- **Datos de productos** (`productos.parquet`): describe las características de los productos del portafolio.
	- `product_id`: identificador único del producto.
	- `brand`: marca comercial del producto.
	- `category`: categoría general del producto, como “BEBIDAS CARBONATADAS”.
	- `sub_category`: subcategoría dentro de la categoría principal, por ejemplo, “GASEOSAS”.
	- `segment`: segmento de mercado al que pertenece el producto, como “PREMIUM”.
	- `package`: tipo de envase del producto.
	- `size`: tamaño del producto en litros.

## 📚 Reglas

<center>
<img src="https://media1.tenor.com/m/0Qtv_cQ4ITsAAAAd/necohaus-grey-name.gif" width="450">



El proyecto consta de **dos entregas parciales** y una **entrega final** en donde la primera entrega la idea es poder reflejar lo aprendido durante la primera mitad del curso, que será sobre los contenidos relacionados a *machine learning*, la segunda será sobre los contenidos de la segunda mitad del curso relacionados a *MLOps* y por último la entrega final constará de dos partes, donde la primera será relacionada con experimentación sobre nuevos datasets que serán disponibilizados durante las últimas semanas del curso de manera incremental y una segunda parte que será el informe final escrito que deberá explicar el desarrollo del proyecto completo, como tambien los resultados y análisis de los experimentos realizados sobre los datasets incrementales. La idea es que todo el código esté desarrollado durante las primeras dos entregas y luego en la entrega final sólo se ejecute el código sobre nuevos conjuntos de datos.

La idea de generar el proyecto por etapas es poder aliviar la carga de trabajo en las últimas semanas del semestre donde sabemos que están muy cargado con entregas, pruebas y exámenes de otros ramos, y así garantizamos que habiendo la desarrollado las dos primeras entregas parciales, tendrán el grueso del proyecto listo para luego experimentar y documentar.

---
### **Fechas de entrega**
- **Entrega parcial 1**: 14 de Mayo
- **Entrega parcial 2**: Por definir
- **Entrega final**: Por definir

---

### **Requisitos del proyecto**
- **Grupos**: Formar equipos de **2 personas**. No se aceptarán trabajos individuales o grupos con más integrantes.
- **Consultas**: Cualquier duda fuera del horario de clases debe ser planteada en el foro correspondiente. Los mensajes enviados al equipo docente serán respondidos únicamente por este medio. Por favor, revisen las respuestas anteriores en el foro antes de realizar nuevas consultas.
- **Plagio**: La copia o reutilización no autorizada de trabajos de otros grupos está **estrictamente prohibida**. El incumplimiento de esta norma implicará la anulación inmediata del proyecto y una posible sanción académica.
- **Material permitido**: Pueden usar cualquier material del curso, ya sea notas, lecturas, códigos, o referencias proporcionadas por los docentes, que consideren útil para el desarrollo del proyecto.

---

### **Entregables y etapas**

#### **1. Entrega Parcial 1**  
- Dispondrán de los archivos de datos **productos.parquet**, **clientes.parquet** y **transacciones.parquet** para el modelamiento inicial.  
- Utilizarán estos archivos para desarrollar lo solicitado para la entrega 1. 
- En esta etapa, se espera que apliquen todos los conocimientos aprendidos durante la primera parte del curso relacionados con *machine learning*.
- **Informe**: No se exige un avance del informe en esta etapa, sólo un notebook con su desarrollo actual, pero se **recomienda comenzar** a redactar el informe final en paralelo para disminuir la carga académica en las etapas posteriores.  

#### **2. Entrega Parcial 2**  
- En esta entrega, deberán aplicar los conocimientos aprendidos durante la segunda mitad del curso sobre *MLOps*  
- Se espera que implementen estos conocimientos para desplegar su modelo elegido en la primera entrega y crear *pipelines* automatizados que simulen un entorno productivo.
- **Informe**: similar a la primera etapa, no se exige un avance del informe, pero se **recomienda avanzar con su redacción** para evitar una acumulación de trabajo en la etapa final.  

#### **3. Entrega Final**  
- En la entrega final, deberán realizar dos etapas:
	- La primera etapa es sobre experimentación utilizando datasets incrementales que se irán disponibilizando de manera parcial, para que vayan generando predicciones con su modelo ya desplegado. El objetivo de esta etapa es poder testear su solución *end-to-end* y que vayan analizando los resultados obtenidos a medida que se van agregando más datos.
	- La segunda etapa consiste en redactar un informe final que deberá explicar el desarrollo completo de tu proyecto y un análisis profundo de sus resultados de experimentación. Este informe debera incluir a lo menos las siguientes secciones:
		- Análisis exploratorio de datos  
		- Metodología aplicada  
		- Selección y entrenamiento de modelos  
		- Evaluación de resultados  
		- Optimización de modelos
		- Interpretabilidad
		- Re-entrenamiento
		- Tracking con MLFlow
		- Creación de la aplicación web con Gradio y FastAPI

Es **altamente recomendable** ir redactando el informe en paralelo al desarrollo de los modelos para garantizar que toda la información relevante quede documentada adecuadamente.  

### Nota Final

La calificación final de su proyecto se calculará utilizando la siguiente ponderación: 

$$Nota Final = 0.30 * EntregaParcial1 + 0.40 * EntregaParcial2 + 0.30 * EntregaFinal$$

---

### **Instrucciones importantes**

1. **Formato del informe**:  
   - El informe debe estar integrado dentro de un **Jupyter Notebook**. No es necesario subirlo a una plataforma externa, pero debe cumplir con los siguientes requisitos:  
     - Estructura clara y ordenada.  
     - Código acompañado de explicaciones detalladas.  
     - Resultados presentados de forma visual y analítica.  

2. **Descuento por informes deficientes**:  
   - Cualquier sección del informe que no tenga una explicación adecuada o no respete el formato será penalizada con un descuento en la nota. Esto incluye código sin comentarios o análisis que no sean coherentes con los resultados presentados.
   - Comentarios sin formatear de ChatGPT o herramientas similares serán penalizados (e.g: "Inserta tu modelo acá", etc.)

# 📬 Entrega Parcial 1 (30% del Proyecto)

### 📪 Fecha de Entrega: 14 de Mayo

## 📌 Abstract [0.25 puntos]

<center>
<img src="https://i.redd.it/h5ptnsyabqvd1.gif" width="400" height="300">

En esta sección, deben redactar un Abstract claro y conciso para su proyecto. El Abstract debe responder a las siguientes preguntas clave:

- **Descripción del problema**: ¿Cuál es el objetivo del proyecto? ¿Qué se intenta predecir o analizar?
- **Datos de entrada**: ¿Qué datos tienen disponibles? ¿Cuáles son sus principales características?
- **Métrica de evaluación**: ¿Cómo medirán el desempeño de sus modelos? Expliquen por qué eligieron esta métrica basándose en el análisis exploratorio de los datos.
- **Modelos y transformaciones**: ¿Qué modelos utilizarán y por qué? ¿Qué transformaciones o preprocesamientos aplicaron a los datos?
- **Resultados generales**: ¿El modelo final cumplió con los objetivos del proyecto? ¿Cuáles fueron las conclusiones más importantes?

**Importante**: Escriban esto despues de haber resuelto el resto de la tarea.

> [Escriba aquí su Abstract]

## 📌 Pre-procesamiento [0.5 puntos]

<center>
<img src="https://media0.giphy.com/media/10zsjaH4g0GgmY/giphy.gif?cid=6c09b9523xtlunksc9amikw09zk1bmiqwjqnt70ae82rk877&ep=v1_gifs_search&rid=giphy.gif&ct=g" width="400" height="300">

Tal como en muchos otros problemas de negocio, los datos probablemente deben ser pre procesados antes de aplicar cualquier técnica de analítica. Bajo esa premisa, en esta sección deben desarrollar código que les permita **preparar los datos** de tal forma que les permita resolver el problema planteado. Para esto, pueden aplicar procesamientos como:

- Transformaciones de tipo de dato (str, int, etc)
- Cruce de información
- Eliminación de duplicados
- Filtros de fila y/o columnas

*Hint: ¿Qué forma debería tener la data para resolver un problema de aprendizaje supervisado?*

Todo proceso llevado a cabo debe estar bien documentado y justificado en el informe, explicando el por qué se decidió realizar en funcion de los datos presentados y los objetivos planteados del proyecto.

In [1]:
#Librerias que vamos a usar en la seccion
import pandas as pd
import datetime
import numpy as np

import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import matplotlib.pyplot as plt


from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler, MinMaxScaler, FunctionTransformer

from sklearn.decomposition import PCA
from sklearn.cluster import KMeans
from sklearn.cluster import DBSCAN
import os
os.environ["OMP_NUM_THREADS"] = "1"

from sklearn.linear_model import LinearRegression
from sklearn.metrics import classification_report


In [2]:
#Nuestros datasets
clientes  = pd.read_parquet('clientes.parquet')
productos = pd.read_parquet('productos.parquet')
transacciones = pd.read_parquet('transacciones.parquet')

In [3]:
clientes.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1569 entries, 10705 to 12273
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   customer_id           1569 non-null   int64  
 1   region_id             1569 non-null   int64  
 2   zone_id               1569 non-null   int64  
 3   customer_type         1569 non-null   object 
 4   Y                     1569 non-null   float64
 5   X                     1568 non-null   float64
 6   num_deliver_per_week  1569 non-null   int64  
 7   num_visit_per_week    1569 non-null   int64  
dtypes: float64(2), int64(5), object(1)
memory usage: 110.3+ KB


In [4]:
productos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 971 entries, 0 to 970
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   product_id    971 non-null    int64  
 1   brand         971 non-null    object 
 2   category      971 non-null    object 
 3   sub_category  971 non-null    object 
 4   segment       971 non-null    object 
 5   package       971 non-null    object 
 6   size          971 non-null    float64
dtypes: float64(1), int64(1), object(5)
memory usage: 53.2+ KB


In [5]:
transacciones.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 254936 entries, 124 to 7712548
Data columns (total 5 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   customer_id    254936 non-null  int64         
 1   product_id     254936 non-null  int64         
 2   order_id       254936 non-null  int64         
 3   purchase_date  254936 non-null  datetime64[ns]
 4   items          254936 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(3)
memory usage: 11.7 MB


Lo primero que podemos notar es que tenemos valores nulos en el dataset. Luego la idea ahora es ampliar la base de transacciones, añadiendo la informacion correspondiente al cliente que reliza la transaccion y al producto transado.

In [6]:
df = pd.merge(transacciones,productos,how='left',on='product_id')
df = pd.merge(df,clientes,how='left',on='customer_id')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 254936 entries, 0 to 254935
Data columns (total 18 columns):
 #   Column                Non-Null Count   Dtype         
---  ------                --------------   -----         
 0   customer_id           254936 non-null  int64         
 1   product_id            254936 non-null  int64         
 2   order_id              254936 non-null  int64         
 3   purchase_date         254936 non-null  datetime64[ns]
 4   items                 254936 non-null  float64       
 5   brand                 254936 non-null  object        
 6   category              254936 non-null  object        
 7   sub_category          254936 non-null  object        
 8   segment               254936 non-null  object        
 9   package               254936 non-null  object        
 10  size                  254936 non-null  float64       
 11  region_id             254936 non-null  int64         
 12  zone_id               254936 non-null  int64         
 13 

A continuacion eliminamos registros duplicados

In [7]:
df = df.drop_duplicates()

Ahora que tenemos la base completa nos gustaria hacer un visualizacion de esta data, para saber si es necesario eliminar columnas irrelevantes o realizar algunas transformaciones. Luego dicha visualizacion puede ser complementada con un analisis por quartiles para enriquecer la exploracion.

In [ ]:
#Ocupamos la funcion del lab4
# Se observa la distribución de las distintas variables
columns = df.drop(columns=['customer_id','product_id','order_id']).columns
fig = make_subplots(rows=5, cols=3, subplot_titles=columns)

# Se agregan las distribuciones
for i in range(len(columns)):
    fig.add_trace(go.Histogram(x=df[columns[i]]), row=i//3 + 1, col=i%3 + 1)

fig.update_layout(height=1000, width=1000, title_text="Distribución de variables numéricas")
    
fig.show()

In [9]:
df['size'].unique()

array([ 1.  ,  0.31,  0.66,  0.25,  0.33, 20.  ,  0.5 ])

In [10]:
df.drop(columns=['customer_id','product_id','order_id']).describe()

items           size  region_id   zone_id              Y  \
count  254051.000000  254051.000000   254051.0  254051.0  254051.000000   
mean        4.375755       0.675775       80.0    5148.0     -46.641522   
std        12.080519       1.088911        0.0       0.0       2.029226   
min      -399.666667       0.250000       80.0    5148.0    -109.002508   
25%         1.000000       0.250000       80.0    5148.0     -46.577054   
50%         2.333333       0.660000       80.0    5148.0     -46.552598   
75%         3.666667       1.000000       80.0    5148.0     -46.489940   
max      1000.333333      20.000000       80.0    5148.0     -46.308598   

                   X  num_deliver_per_week  num_visit_per_week  
count  254051.000000         254051.000000            254051.0  
mean     -107.777588              3.164680                 1.0  
std         2.030981              0.392777                 0.0  
min      -108.620920              2.000000                 1.0  
25%      -107.930286              3.000000                 1.0  
50%      -107.892716              3.000000                 1.0  
75%      -107.856730              3.000000                 1.0  
max       -46.442785              6.000000                 1.0

Hagamos un analisis por cada variable:

* **purchase date**: Se observa una cierta estacionalidad con tendencia en la cantidad de transacciones realizadas por fecha

* **items**: La cantidad de items se ve concentrada entre 1 y 3 items sin embargo hay claras señales de outliers que se pueden ver tanto en el grafico de distribuciones como en los quartiles

* **brand**: La gran mayoria de transacciones se concentran en 6 marcas de un total de 21 marcas

* **category**: Todas las transacciones pertenecen a la categoria de bebidas carbonatadas, por tanto esta no es una variable relevante para generar predicciones posteriormente. 

* **sub category**: Solo tenemos 2 subgcaegorias y la gran mayoria de transacciones pertenecen a gaseosas, aguas saborizadas solo representan el 0,17% de las transacciones.

* **segment**: Los segmentos se ven liderados por la categoria medium, luego el resta de categorias son relativamente uniformes en distribucion. 

* **package**: Tenemos 3 categorias y practicamente todas las transacciones estan concentradas en botellas y latas, Keg solo representa el 0.3% de las transacciones.

* **size**: La gran mayoria de las transacciones estan concentradas en el rango [0.25,1], pero tal como se puede ver en los rangos quartilicos y en el grafico de distribucion, hay outliers que no nos permiten obsvervar correctamente la distribucion. Sin embargo a pesar de ser una variable numerica, solo existen 7 tipos de tamaños lo que nos invita a tratar esta variable como una categorica.

* **region id**: Todas las transacciones corresponden a la misma region, que es la region 80. Es por esta razon que para efecto de las predicciones no es una variable relevante. 

* **zone id**: Al igual que la region, todas las transacciones son realizadas en la misma zona (5148). No es una variable relevante para realizar predicciones. 

* **customer type**: Tenemos 7 categorias, luego gran parte de las transacciones corresponden a abarrotes. 

* **X,Y**: Referentes a las coordenadas de la ubicacion de la tienda. Todas las tiendas parecen concentrarse en una region determinada ([-46.5,-46.3],[-108.6,-107.8]), por la naturaleza de los outliers en cada columna, de la impresion de que los outliers son errores de typeo donde las coordenadas X e Y estan invertidas (Posteriormente podemos hacer un analisis para confirmar esta hipotesis).

* **num deliver per week**: El numero de entregas por semana se encuentra en el rango [2,6], donde la mayoria de los clientes asociadas a las transacciones reciben 3 entregas

* **num visit per week**: Todos los clientes asociados a las transacciones tienen 1 visita semanal a la fuerza de venta, por tanto no es una varible relevante a la hora de hacer predicciones.

Una vez hecho este analisis procedemos a eliminar las variables que no nos resultan relevantes a la hora de hacer predicciones



In [11]:
df.drop(columns=['category','region_id','zone_id','num_visit_per_week'], inplace=True)

## 📌 EDA [0.5 puntos]

<center>
<img src="https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExbHZ6aGdkd21tYTI3cW8zYWhyYW5wdGlyb2s3MmRzeTV0dzQ1NWlueiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/3k1hJubTtOAKPKx4k3/giphy.gif" width="400" height="200">

En esta sección, se debe realizar un análisis exploratorio de los datos para comprender su estructura, detectar posibles problemas y obtener información relevante para el entrenamiento de los modelos. La idea es que puedan detectar **patrones en los datos** que les permitan resolver el problema con mayor facilidad.

Se deben responder preguntas a partir de lo que puedan visualizar/obtener, por ejemplo:

- Clientes y productos

    - ¿Cuántos clientes únicos hay en el dataset?

    - ¿Cuántos productos únicos se encuentran en los datos?

- Periodo y frecuencia

    - ¿De qué periodo es la información disponible?

    - ¿Cuál es la frecuencia de los registros (diaria, semanal, mensual, etc.)?

- Calidad de los datos

    - ¿Existen valores nulos en el dataset? ¿Cuántos? ¿Cómo se pueden tratar?

    - ¿Hay datos raros, como cantidades negativas o inconsistencias? Genere tests de validación para identificar estos problemas.

- Patrones de compra

    - ¿Cuántos productos compra en promedio cada cliente semana a semana?

    - ¿Cuántas transacciones ha realizado cada cliente?

    - ¿Cuál es el periodo de recompra promedio de cada SKU?

### Clientes, productos y periodo

In [12]:
clientes_unicos = df['customer_id'].nunique()
productos_unicos = df['product_id'].nunique()
fecha_min = df['purchase_date'].min().strftime('%d/%m/%Y') #Esto es para que no salga la hora en la tabla
fecha_max = df['purchase_date'].max().strftime('%d/%m/%Y')
periodo = f"{fecha_min} - {fecha_max}"

fig = go.Figure(data=[go.Table(
    header=dict(
        values=["Métrica", "Valor"],
        fill_color='paleturquoise',
        align='left',
        font=dict(size=14, color='black')
    ),
    cells=dict(
        values=[
            ["Clientes únicos", "Productos únicos", "Período de registros"],
            [clientes_unicos, productos_unicos, periodo]
        ],
        fill_color='lavender',
        align='left',
        font=dict(size=13)
    )
)])

fig.update_layout(title="Resumen de los datos", width=600, height=300)
fig.show()

De la tabla anterior podemos ver en primer lugar que no todos los clientes en nuestra base de datos hacen transacciones y que no todos los productos en nuestro catalogo son transados, pues en nuestra base de datos hay 1569 clientes y 971 productos. Luego el periodo de transacciones observado es de 1 año. En cuanto a lo anterior es razonable querer añadir estos clientes - productos pues estos productos - clientes pueden compartir muchas similitudes con los que efectivamente se ve en el dateset. Esto generaría que el modelo asocie esos productos a no compra aun cuando si lo pueden ser. Y como la gran mayoria de los productos no son transados las probabilidades disminuiran mucho. 

### Frecuencia

In [ ]:
# Agrupaciones por frecuencia
por_dia = df.groupby(df['purchase_date'].dt.date).size()
por_semana = df.groupby(pd.Grouper(key='purchase_date', freq='W')).size()
por_mes = df.groupby(pd.Grouper(key='purchase_date', freq='M')).size()

# los 3 subplots
fig = make_subplots(rows=3, cols=1, shared_xaxes=False,
                    subplot_titles=("Frecuencia diaria", "Frecuencia semanal", "Frecuencia mensual"))

# Histograma diario
fig.add_trace(go.Bar(x=por_dia.index, y=por_dia.values, name='Diario'),
              row=1, col=1)

# Histograma semanal
fig.add_trace(go.Bar(x=por_semana.index, y=por_semana.values, name='Semanal'),
              row=2, col=1)

# Histograma mensual
fig.add_trace(go.Bar(x=por_mes.index, y=por_mes.values, name='Mensual'),
              row=3, col=1)

fig.update_layout(height=900, width=1000, title_text="Frecuencia de registros por día, semana y mes")
fig.update_xaxes(title_text="Fecha", row=3, col=1)
fig.update_yaxes(title_text="Cantidad de registros", row=1, col=1)
fig.update_yaxes(title_text="Cantidad de registros", row=2, col=1)
fig.update_yaxes(title_text="Cantidad de registros", row=3, col=1)

fig.show()

Lo primero que podemos analizar del grafico anterior es que hay estacionalidad semanal, es decir hay dias de la semana en particular donde probablemente se hacen los pedidos, luego cuando vemos la frecuencia a nivel de semana y mes es claro ver que hay periodos de tiempo en particular (Junio Y Agosto) donde las transacciones caen bastante, y periodos previos a navidad - año nuevo donde las transacciones aumentan mucho.

### Calidad de los datos

Como vimos en la seccion anterior no hay valores faltantes, sin embargo si hay anomalias en los datos, en *items* y *size* hay valores atipicos mientras que en las coordenadas X e Y puden haber errores de typeo u outliers, en el caso de *items* y *size* podemos hacer una normalizacion, estandarizacion de los datos o eliminarlos. Mientras que para las coordenadas X e Y podemos hacer un analisis de aquellas tiendas que tienen valores atipicos para ver a que cliente en particular pertenencen y discernir si corresponden a ubicaciones reales o errores de typeo. Finalmente mencionar que en *items* no solo tenemos valores atipicos, si no que tambien negativos, esto no es tan extraño pues los valores negativos pueden corresponder a devoluciones, este es un punto sumamente importante pues el objetivo del proyecto es predecir probabilidades de compra. ¿Nos servirá un cliente que compra 100 y devuelve 90? Mas aun, ¿Queremos ofrecerle 100 a alguien que nos devolverá 90? Quizas lo mejor sería que ese cliente se considere como uno que simplemente compra 10. Lo que sugiere este razonamiento es agrupar las compras y devoluciones de un mismo producto para cada cliente.

Como vimos en la seccion anterior no tenemos valores faltantes en nuestro dataset, sin embargo no olvidemos que el objetivo de este proyecto es hacer un sistema de recomendacion basado en probabilidades de compra, cual es nuestra variable a predecir y? Son nuestros registros, en efecto cada registro corresponde a una compra y los registros que no corresponden a compras son los registros que no existen en nuestro datset. Como hemos visto en la parte anterior tenemos 1490 tiendas, 114 clientes y 366 dias. Como nosotros queremos estudiar la probabilidad de que un cliente compre un determinado producto en una fecha dada, necesitamos la combinacion de cliente, producto, dia para cada combinacion de estos valores. Esto nos da un total de 62.168.760 registros y sin embargo tenemos tan solo 254.051. Esto quiere decir que tenemos 61.914.709 registros en donde no hay compras. Si hacemos la ampliacion del dataset estos registros donde no hay compras tendrian en sus columnas *order_id, items* valores nulos. Que hacemos con estos valores nulos? A la columna items le asignamos el valor 0, pues decir que no existió la transaccion es equivalente a decir que se transaron 0 items. Por otra parte podemos asigar un unico order_id a todos los registros inexistentes. Finalmente agregar que queremos hacer predicciones semanales, por tanto se deben agregar los datos a esta temporalidad.

In [14]:
df.columns

Index(['customer_id', 'product_id', 'order_id', 'purchase_date', 'items',
       'brand', 'sub_category', 'segment', 'package', 'size', 'customer_type',
       'Y', 'X', 'num_deliver_per_week'],
      dtype='object')

In [15]:
#  variable de semana (año-semana) por si despues entregan otro año
transacciones['year_week'] = transacciones['purchase_date'].dt.strftime('%Y-%U')  # Semana del año

# agregación semanal
cols_to_keep = ['customer_id', 'product_id', 'year_week']


weekly_df = transacciones.groupby(cols_to_keep).agg({
    'items': 'sum',
}).reset_index()

# los valores únicos observados en las transacciones y que por tanto hay que sacar de df y no de las BD originales
customers = df['customer_id'].unique()
products = df['product_id'].unique()
weeks = df['purchase_date'].dt.strftime('%Y-%U').unique()

# combinaciones posibles cliente-producto-semana
full_index = pd.MultiIndex.from_product(
    [customers, products, weeks],
    names=['customer_id', 'product_id', 'year_week']
)
full_df = pd.DataFrame(index=full_index).reset_index()

# se une con el dataset agregado
df_completo = pd.merge(
    full_df,
    weekly_df,
    on=['customer_id', 'product_id', 'year_week'],
    how='left'
)

# se completa con las variables de clientes y productos
df_completo = df_completo.merge(clientes, on='customer_id', how='left')
df_completo = df_completo.merge(productos, on='product_id', how='left')

df_completo['items'] = df_completo['items'].fillna(0)

# Nuestra variable de interes, aqui en particular puede que haya existido un registro y que tenga items 0
# aqui lo estamos considerando como no compra
df_completo['y'] = (df_completo['items'] > 0).astype(int)

#Finalemente esto es para definir la variable temporal como una de tipo datetime y no un string
df_completo['year_week'] = pd.to_datetime(df_completo['year_week'] + '-1', format='%Y-%U-%w')


In [16]:
df_completo.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9002580 entries, 0 to 9002579
Data columns (total 18 columns):
 #   Column                Dtype         
---  ------                -----         
 0   customer_id           int64         
 1   product_id            int64         
 2   year_week             datetime64[ns]
 3   items                 float64       
 4   region_id             int64         
 5   zone_id               int64         
 6   customer_type         object        
 7   Y                     float64       
 8   X                     float64       
 9   num_deliver_per_week  int64         
 10  num_visit_per_week    int64         
 11  brand                 object        
 12  category              object        
 13  sub_category          object        
 14  segment               object        
 15  package               object        
 16  size                  float64       
 17  y                     int32         
dtypes: datetime64[ns](1), float64(4), int32(1)

### Patrones de compra

In [ ]:
# promedio de productos por semana por cliente
promedio_productos_semana = (
    df_completo.groupby('customer_id')['items'].mean().reset_index(name='prom_produc_semana')
)

df_promedio_productos = promedio_productos_semana.reset_index()

# transacciones por cliente
transacciones = df[['customer_id', 'order_id']].drop_duplicates()
transacciones_por_cliente = (
    transacciones.groupby('customer_id').size().reset_index(name='transaction_count')
)

df_transacciones_cliente = transacciones_por_cliente.reset_index()

#Añadimos esta features a nuestro dataframe
df_completo = df_completo.merge(promedio_productos_semana, on='customer_id', how='left')
df_completo = df_completo.merge(transacciones_por_cliente, on='customer_id', how='left')


# la subfigura
fig = make_subplots(
    rows=1, cols=2,
    subplot_titles=('Promedio de productos por semana', 'Transacciones por cliente'),
)

# Gráfico para "Promedio de productos por semana"
fig.add_trace(go.Bar(
    x=df_promedio_productos.index,
    y=df_promedio_productos['prom_produc_semana'],
    marker=dict(color='skyblue', line=dict(width=2, color='lightblue')),
    name='Promedio de productos',
), row=1, col=1)

# Gráfico para "Número de transacciones por cliente"
fig.add_trace(go.Bar(
    x=df_transacciones_cliente.index, 
    y=df_transacciones_cliente['transaction_count'], 
    marker=dict(color='lightcoral', line=dict(width=2, color='lightblue')),
    name='Transacciones por cliente',
), row=1, col=2)

fig.update_layout(
    title_text="Promedio de productos y transacciones por cliente",
    plot_bgcolor='white',  
    yaxis_title='Promedio de productos',
    xaxis=dict(tickangle=45),
    yaxis=dict(title='Promedio de productos'),  
    yaxis2=dict(title='Número de transacciones', 
                overlaying='y',  
                side='right'), 
    bargap=0.1,  
    showlegend=False 
)

fig.show()

De los graficos anteriores se puede ver en primer lugar que la mayoria de los clientes tiene un comportamiento parecido en la cantidad de productos que compran semanalmente, a excepcion de un par de clientes que se pueden estar viendo influenciados por los outliers que vimos en la seccion anterior, luego en cuanto a la cantidad de transacciones realizadas la mayor cantidad de clientes se comportan de manera similar sin embargo se pueden distinguir dos grupos, los que realizan alrededor de 50 transacciones y  los que realizan 100 transacciones

In [ ]:
df['purchase_date'] = pd.to_datetime(df['purchase_date'])

df_sorted = df.sort_values(['product_id', 'purchase_date'])

df_sorted['recompra_promedio'] = df_sorted.groupby('product_id')['purchase_date'].diff().dt.days

recompra_promedio = df_sorted.groupby('product_id')['recompra_promedio'].mean()

#Añadimos la variable al dataset
df_completo = df_completo.merge(recompra_promedio, on='product_id', how='left')

recompra_promedio_reset = recompra_promedio.reset_index()

fig = make_subplots(
    rows=1, cols=1, 
    subplot_titles=('Recompra promedio por SKU',)
)

fig.add_trace(go.Bar(
    x=recompra_promedio_reset.index, 
    y=recompra_promedio_reset['recompra_promedio'],
    marker=dict(color='lightblue', line=dict(width=2, color='lightblue')),
    name='Recompra promedio',
), row=1, col=1)

fig.update_layout(
    title_text="Recompra promedio por SKU",
    plot_bgcolor='white', 
    xaxis_title='Índice de producto',
    yaxis_title='Promedio de días entre compras', 
    xaxis=dict(tickangle=45),
    yaxis=dict(title='Promedio de días entre compras'),  
    bargap=0.1, 
    showlegend=False 
)

fig.show()




En este grafico se puede ver que la mayoria de los productos tienen una alta rotacion, a excepcion de un par que no se compran mucho.

In [19]:
df_completo.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9002580 entries, 0 to 9002579
Data columns (total 21 columns):
 #   Column                Dtype         
---  ------                -----         
 0   customer_id           int64         
 1   product_id            int64         
 2   year_week             datetime64[ns]
 3   items                 float64       
 4   region_id             int64         
 5   zone_id               int64         
 6   customer_type         object        
 7   Y                     float64       
 8   X                     float64       
 9   num_deliver_per_week  int64         
 10  num_visit_per_week    int64         
 11  brand                 object        
 12  category              object        
 13  sub_category          object        
 14  segment               object        
 15  package               object        
 16  size                  float64       
 17  y                     int32         
 18  prom_produc_semana    float64       
 19  

## 📌 Holdout [0.25 puntos]

Para evaluar correctamente el modelo y garantizar su capacidad de generalización, se deben dividir los datos en tres conjuntos: 
- `Entrenamiento` : Para ajustar los parámetros.
- `Validación`: Para optimizar hiperparámetros y seleccionar el mejor modelo.
- `Prueba` : Para evaluar el rendimiento final en datos no vistos.

👀 **Hint**: *Recuerde que los datos tienen una temporalidad que debe considerarse al momento de separarlos, para evitar fugas de información. Es importante justificar la estrategia de partición elegida y visualizar la distribución temporal de los conjuntos generados*

Aqui la estrategia es no usar un train test split aleatorio, pues estariamos mezclando eventos pasados con eventos futuros, por tanto la idea es ordenar por fecha el df y luego dividir.

In [20]:
df_completo = df_completo.sort_values('year_week')

n = len(df_completo)
train_end = int(n * 0.6)
val_end = int(n * 0.8)

df_train = df_completo.iloc[:train_end]
df_val = df_completo.iloc[train_end:val_end]
df_test = df_completo.iloc[val_end:]
df_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 5401548 entries, 6776133 to 5292012
Data columns (total 21 columns):
 #   Column                Dtype         
---  ------                -----         
 0   customer_id           int64         
 1   product_id            int64         
 2   year_week             datetime64[ns]
 3   items                 float64       
 4   region_id             int64         
 5   zone_id               int64         
 6   customer_type         object        
 7   Y                     float64       
 8   X                     float64       
 9   num_deliver_per_week  int64         
 10  num_visit_per_week    int64         
 11  brand                 object        
 12  category              object        
 13  sub_category          object        
 14  segment               object        
 15  package               object        
 16  size                  float64       
 17  y                     int32         
 18  prom_produc_semana    float64       

In [ ]:
fecha_entrenamiento = df_train['year_week'].max().to_pydatetime()
fecha_validacion = df_val['year_week'].max().to_pydatetime()

fig = go.Figure()

fig.add_trace(go.Scatter(
    x=df_completo['year_week'],
    y=df_completo['items'],
    mode='lines',
    name='Total',
    line=dict(color='blue'),
    opacity=0.3
))

fig.add_shape(
    type='line',
    x0=fecha_entrenamiento,
    y0=0,
    x1=fecha_entrenamiento,
    y1=df_completo['items'].max(),
    line=dict(color='green', dash='dash'),
)

fig.add_annotation(
    x=fecha_entrenamiento,
    y=df_completo['items'].max(),
    text="Entrenamiento",
    showarrow=True,
    arrowhead=1,
    ax=0,
    ay=-40
)

fig.add_shape(
    type='line',
    x0=fecha_validacion,
    y0=0,
    x1=fecha_validacion,
    y1=df_completo['items'].max(),
    line=dict(color='orange', dash='dash'),
)

fig.add_annotation(
    x=fecha_validacion,
    y=df_completo['items'].max(),
    text="Validación",
    showarrow=True,
    arrowhead=1,
    ax=0,
    ay=-40
)

fig.update_layout(
    title="Partición temporal de los datos",
    xaxis_title="Fecha",
    yaxis_title="Cantidad de ítems",
    template="plotly_white",
    width=1000,
    height=400
)

fig.show()

## 📌 Feature Engineering [0.5 puntos]

<center>
<img src="https://i.imgur.com/CmXZSSC.gif" width="300" height="200">

En esta sección, se deben construir pipelines para automatizar el preprocesamiento de los datos, lo cual garantizará que el flujo de trabajo sea reproducible y eficiente para esta entrega y las futuras. El objetivo es aplicar una serie de transformaciones en un orden definido para asegurar que los datos estén listos para los modelos a entrenar. El pipeline final debe incluir las técnicas de pre-procesamiento que se deben aplicar a los distintos datos (según lo que consideren necesario para el problema). Por ejemplo:

- **Imputación de valores nulos**: Manejo de datos faltantes mediante estrategias adecuadas (media, mediana, moda, interpolación, etc.). 

- **Transformaciones personalizadas**: Uso de ColumnTransformer para aplicar diferentes transformaciones a columnas específicas.

- **Codificación de variables categóricas**: Convertir datos categóricos a un formato numérico adecuado (One-Hot Encoding, Label Encoding, etc.).

- **Discretización de variables**: Conversión de variables numéricas continuas en categorías si son relevantes para el desempeño del modelo a entrenar.

- **Estandarización o normalización** : Ajustar la escala de los datos para mejorar el rendimiento de los algoritmos sensibles a la magnitud de las variables.

- **Eliminación o transformación de valores atípicos**: Identificar y tratar con datos outliers para mejorar la robustez del modelo.

- **Nuevas características** : Creación de variables adicionales que puedan aportar información relevante al modelo.

Cada una de estas transformaciones debe ser justificada en función de su relevancia para el problema y los datos, y es importante evaluar su impacto en el rendimiento del modelo. Además, el pipeline debe ser flexible y modular para poder probar diferentes configuraciones de preprocesamiento.

### Purchase Date

In [22]:
class DateParser(BaseEstimator, TransformerMixin):
    def fit(self, X, y=None):
        return self

    def transform(self, X):
        fecha = pd.to_datetime(X.iloc[:, 0], format='%Y-%U-%w')
        return pd.DataFrame({
            'week': fecha.dt.isocalendar().week.astype('int'),
            'month': fecha.dt.month.astype('int'), #igual añadimos el mes
            'year': fecha.dt.year.astype('int')
        }, index=X.index)

    def set_output(self, transform='default'):
        return self

### Size

Como vimos anteriormente si bien esta variable es numerica, es de naturaleza categorica, pues clasifica los bebestibles en 7 tamaños distintos, para hacer un tratamiento de este podemos crear una clase que transforme esta variable en categorica, para luego pasarla a traves de un encoder. Con la exploracion realizada en la seccion de pre-procesamiento sabemos que hay 7 categorias de tamaños, sin embargo se pueden resumir en 4: pequeñas : $[0.25,0.31,0.33]$; medianas : $[0.5,0.66]$; grandes: $[1]$; muy grandes: $[20]$.

In [23]:
class CategoricaSize(BaseEstimator, TransformerMixin):
    def __init__(self, col='size'):
        self.col = col

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        X = X.copy()

        def categorizar(val):
            if val in [0.25, 0.31, 0.33]:
                return 'pequeña'
            elif val in [0.5, 0.66]:
                return 'mediana'
            elif val == 1:
                return 'grande'
            elif val == 20:
                return 'muy_grande'


        X[self.col] = X[self.col].apply(categorizar).astype('category')
        return X
    
    def set_output(self,transform='default'):
        return self

### X e Y

Ahora hagamos el tratamiento de las coordenadas X y Y mencionadas anteriormente

In [24]:
df[df['X'] > -100]['customer_id'].unique()

array([236766, 219231, 165126], dtype=int64)

In [25]:
df[df['Y'] < -50]['customer_id'].unique()

array([236766, 219231, 165126], dtype=int64)

Lo primero aqui es observar que los outliers que vimos en la coordenadas X e Y pertenecen a los mismos clientes

In [26]:
df[df['customer_id'] == 236766]['Y'].unique()

array([-107.879946])

In [27]:
df[df['customer_id'] == 219231]['Y'].unique()

array([-109.0025078])

In [28]:
df[df['customer_id'] == 165126]['Y'].unique()

array([-107.899346])

Como podemos ver todas las tiendas asociadas a esos registros tienen la misma ubicacion, por tanto no podemos ver que existe un ejemplo dentro de esas tiendas donde se vea que las coordenadas estan invertidas. Sin embargo hay algo que si sabemos y es que todas las tiendas se ubican en la misma region y en la misma zona, por tanto debiesen estar en lugares cercanos. Veamos cuanta distancia hay entre los puntos que suponemos que tienen coordenadas invertidas y el resto de los otros puntos para ver si son puntos factibles

In [29]:
# Función Haversine
def haversine(lat1, lon1, lat2, lon2):
    R = 6371  # Radio de la Tierra en km 
    phi1, phi2 = np.radians(lat1), np.radians(lat2)
    delta_phi = np.radians(lat2 - lat1)
    delta_lambda = np.radians(lon2 - lon1)
    
    a = np.sin(delta_phi / 2.0)**2 + \
        np.cos(phi1) * np.cos(phi2) * np.sin(delta_lambda / 2.0)**2
    c = 2 * np.arcsin(np.sqrt(a))
    
    return R * c

# Una tienda normal
x1, y1 = -107.125965, -46.489646  # X: longitud, Y: latitud

# Una tienda con coordenadas supuestamente invertidas
x2, y2 = -46.589622, -107.899346

dist_km = haversine(y1, x1, y2, x2)
print(f"Distancia entre puntos: {dist_km:.2f} km")

Distancia entre puntos: 6018.06 km


Una distancia de 6000 km es mas que la distancia que hay desde Arica hasta Puerto Willams, por tanto en vista de que sabemos que ambos puntos se encuentran en la misma zona, podemos concluir que los puntos estan invertidos. Creemos una funcion que nos permita tener correctamente las variables y luego veamos como se ven en un plano

In [ ]:
def invertir_coordenadas(df_completo, x_col='X', y_col='Y'):
    df_completo = df_completo.copy()

    x_valido = df_completo[x_col].between(-110, -106)
    y_valido = df_completo[y_col].between(-47, -45)
    fuera_de_rango = ~(x_valido & y_valido)

    x_invertido_valido = df_completo[y_col].between(-110, -106)
    y_invertido_valido = df_completo[x_col].between(-47, -45)
    se_pueden_invertir = fuera_de_rango & x_invertido_valido & y_invertido_valido

    df_completo.loc[se_pueden_invertir, [x_col, y_col]] = df_completo.loc[se_pueden_invertir, [y_col, x_col]].values

    return df_completo

df_completo = invertir_coordenadas(df_completo)

plt.figure(figsize=(10, 5))

plt.subplot(1, 1, 1)
plt.scatter(df_completo['X'], df_completo['Y'], alpha=0.7, c='green', label='Corregido')
plt.title('Después de invertir')
plt.xlabel('X')
plt.ylabel('Y')
plt.grid(True)
plt.legend()

plt.tight_layout()
plt.show()


Se puede ver de forma mucho mas clara la distribucion de las tiendas en el plano. Ahora la implementacion en una clase

In [31]:
class InvertirTransformer(BaseEstimator, TransformerMixin):
    def __init__(self, x_col='X', y_col='Y'):
        self.x_col = x_col
        self.y_col = y_col

    def fit(self, X, y=None):
        return self

    def transform(self, X):
        X = X.copy()

        x_valido = X[self.x_col].between(-110, -106)
        y_valido = X[self.y_col].between(-47, -45)
        fuera_de_rango = ~(x_valido & y_valido)

        x_invertido_valido = X[self.y_col].between(-110, -106)
        y_invertido_valido = X[self.x_col].between(-47, -45)
        se_pueden_invertir = fuera_de_rango & x_invertido_valido & y_invertido_valido

        X.loc[se_pueden_invertir, [self.x_col, self.y_col]] = X.loc[se_pueden_invertir, [self.y_col, self.x_col]].values

        return X
    
    def set_output(self,transform='default'):
        return self



### Z-score

Como hemos visualizado en la seccion de preprocesamiento, la gran cantidad de items se concentra en valores bajos en efecto mas del 75% de las transacciones tienen asociados menos de 4 items. Sin embargo tenemos transacciones con 200, 300 hasta 1000 items, es razonable pensar que estos no son valores reales y que por tanto deben ser eliminados.

In [32]:
class ZScoreFilter(BaseEstimator, TransformerMixin):
    def __init__(self, col='items', threshold=1.96):
        self.col = col
        self.threshold = threshold

    def fit(self, X, y=None):
        self.mean_ = X[self.col].mean()
        self.std_ = X[self.col].std()
        return self

    def transform(self, X):
        X = X.copy()
        z_scores = (X[self.col] - self.mean_) / self.std_
        mask = np.abs(z_scores) <= self.threshold
        return X[mask].reset_index(drop=True)
    
    def set_output(self,transform='default'):
        return self


### Brand cluster

In [33]:
df['brand'].nunique()

21

Al ser la marca una variable categorica, para efectos del modelo tendremos que codificar como una dummy cad uno de sus valores (21). Para reducir la dimensionalidad de las marcas podemos hacer un analisis de clustering sobre las marcas y agrupar aquellas que son parecidas. Primero veamos la eleccion de los clusters y luego una implementacion mas formal para incluirla en el pipeline general

In [ ]:
# columnas relevantes
df_cluster = df_completo[['brand', 'segment','package','size']].copy()

# eliminamos duplicados por producto para evitar sesgo por transacción
df_cluster = df_cluster.drop_duplicates(subset=['brand', 'segment', 'package','size'])

# brand
agg = df_cluster.groupby('brand').agg({
    'size' : 'mean',
    'package': lambda x: x.mode()[0],
    'segment': lambda x: x.mode()[0],
}).reset_index()

# One-hot encoding para variables categóricas
ohe = OneHotEncoder(sparse_output=False)
encoded = ohe.fit_transform(agg[['segment','package']])
encoded_df = pd.DataFrame(encoded, columns=ohe.get_feature_names_out(['segment','package']))
encoded_df.index = agg.index  # Alinear índices para concatenación

# Construcción del dataset para clustering
X = pd.concat([agg[['size']], encoded_df], axis=1)

# Clustering con KMeans
kmeans = KMeans(n_clusters=3, random_state=0)
agg['cluster'] = kmeans.fit_predict(X)

# PCA
pca = PCA(n_components=2)
X_2d = pca.fit_transform(X)

plt.figure(figsize=(8, 6))
for cluster in agg['cluster'].unique():
    plt.scatter(
        X_2d[agg['cluster'] == cluster, 0],
        X_2d[agg['cluster'] == cluster, 1],
        label=f'Cluster {cluster}'
    )

for i, brand in enumerate(agg['brand']):
    plt.text(X_2d[i, 0], X_2d[i, 1], brand, fontsize=8)

plt.title("Clusters de Marcas (Brand) con `items` incluido")
plt.xlabel("PCA 1")
plt.ylabel("PCA 2")
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()


Hay un cumulo de tiendas en el sector inferior izquierdo, sumado a pequeños grupos de tiendas que parecen estar cerca en la esquina superior izquierda, si bien no son clusters claramente definidos tener 3 tiendas relativamente parecidas es mejor que 21 marcas por separado. Ahora podemos hacer la transformacion a traves de una clase para incorporarlo al pipeline

In [52]:
#Idem a la parte anterior pero en la clase
class BrandClusterTransformer(BaseEstimator, TransformerMixin):
    def __init__(self, n_clusters=3):
        self.n_clusters = n_clusters
        self.encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
        self.scaler = StandardScaler()
        self.kmeans = KMeans(n_clusters=self.n_clusters, random_state=0)
    
    def fit(self, X, y=None):
        df = X.copy()

        df = df.drop_duplicates(subset=['brand', 'segment', 'size','package'])

        # Agrupación por brand
        agg = df.groupby('brand').agg({
            'size' : 'mean',
            'package': lambda x: x.mode()[0],
            'segment': lambda x: x.mode()[0] if not x.mode().empty else 'otro'
        }).reset_index()

        self.agg_ = agg.copy()

        segment_encoded = self.encoder.fit_transform(agg[['segment','package']])
        segment_df = pd.DataFrame(segment_encoded, columns=self.encoder.get_feature_names_out(['segment','package']))
        segment_df.index = agg.index

        X_cluster = pd.concat([agg[['size']], segment_df], axis=1)

        X_scaled = self.scaler.fit_transform(X_cluster)

        self.kmeans.fit(X_scaled)

        self.brand_to_cluster = dict(zip(agg['brand'], self.kmeans.labels_))

        return self
    
    def transform(self, X):
        X = X.copy()
        X['brand_cluster'] = X['brand'].map(self.brand_to_cluster)
        return X
    
    def set_output(self, transform='default'):
        return self


### X e Y cluster

In [ ]:
#Utilizamos las variables relacionadas al cliente - tienda para hacer el cluster
df_cluster = df_completo[['X', 'Y', 'customer_type', 'num_deliver_per_week', 'num_visit_per_week']].copy()

agg = df_cluster.groupby(['X', 'Y']).agg({
    'customer_type': lambda x: x.mode()[0] if not x.mode().empty else 'otro',
    'num_deliver_per_week': 'mean',
    'num_visit_per_week': 'mean'
}).reset_index()

ohe = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
customer_encoded = ohe.fit_transform(agg[['customer_type']])
customer_encoded_df = pd.DataFrame(customer_encoded, columns=ohe.get_feature_names_out(['customer_type']))

X = pd.concat([
    agg[['num_deliver_per_week', 'num_visit_per_week']].reset_index(drop=True),
    customer_encoded_df.reset_index(drop=True)
], axis=1)

# Aplicación de DBSCAN
dbscan = DBSCAN(eps=0.7, min_samples=80)
agg['cluster'] = dbscan.fit_predict(X)

# PCA
pca = PCA(n_components=2)
X_2d = pca.fit_transform(X)

plt.figure(figsize=(8, 6))
for cluster in agg['cluster'].unique():
    plt.scatter(
        X_2d[agg['cluster'] == cluster, 0],
        X_2d[agg['cluster'] == cluster, 1],
        label=f'Cluster {cluster}'
    )

for i, (x, y) in enumerate(X_2d):
    plt.text(x, y, f'{agg["X"].iloc[i]}, {agg["Y"].iloc[i]}', fontsize=8)

plt.title("Visualización de Clusters (X, Y, items, size)")
plt.xlabel("PCA 1")
plt.ylabel("PCA 2")
plt.legend()
plt.grid(True)
plt.tight_layout()
plt.show()

Utilizando una tecnica de clustering basada en densidad podemos encontrar 2 clusters mas un par outliers. Ahora la implementacion en una clase

In [37]:
#Lo mismo pero en la clase
class XYClusterTransformer(BaseEstimator, TransformerMixin):
    def __init__(self, eps=0.7, min_samples=80):
        self.eps = eps
        self.min_samples = min_samples
        self.encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')
        self.dbscan = DBSCAN(eps=self.eps, min_samples=self.min_samples)

    def fit(self, X, y=None):
        df = X.copy()

        agg = df.groupby(['X', 'Y']).agg({
            'customer_type': lambda x: x.mode()[0] if not x.mode().empty else 'otro',
            'num_deliver_per_week': 'mean',
            'num_visit_per_week': 'mean'
        }).reset_index()

        customer_encoded = self.encoder.fit_transform(agg[['customer_type']])
        X_cluster = np.concatenate([
            agg[['num_deliver_per_week', 'num_visit_per_week']].values,
            customer_encoded
        ], axis=1)

        self.dbscan.fit(X_cluster)
        self.XY_to_cluster = dict(zip(zip(agg['X'], agg['Y']), self.dbscan.labels_))
        return self

    def transform(self, X):
        X = X.copy()
        X['XY_cluster'] = X.apply(lambda row: self.XY_to_cluster.get((row['X'], row['Y']), -1), axis=1)
        return X

    def set_output(self, transform='default'):
        return self

### PipeLine inicial

In [53]:
from sklearn.impute import SimpleImputer
lista_stand = ['num_deliver_per_week','transaction_count','recompra_promedio']
lista_categoricas = ['package', 'customer_type', 'segment','brand_cluster','XY_cluster']
lista_temporales = ['year_week']
lista_invertir = ['X','Y']
lista_convertir = ['size']


trans_stand = Pipeline([
    ('simple_imputer', SimpleImputer(strategy='mean')), 
    ('MinMaxScaler', MinMaxScaler())
])

trans_categoricas = Pipeline([
    ('OneHotEncoder', OneHotEncoder(sparse_output=False, handle_unknown='ignore'))
])

trans_temporales = Pipeline([
    ('Date_Parser', DateParser())
])

trans_invertir = Pipeline([
    ('Invertir', InvertirTransformer())
])

trans_convertir = Pipeline([
    ('Convertir',CategoricaSize()),
    ('OneHotEncoder_size', OneHotEncoder(sparse_output=False, handle_unknown='ignore'))
])

preprocesamiento = Pipeline([
    ('inv', InvertirTransformer()),
    ('brand_cluster', BrandClusterTransformer(n_clusters=3)),
    ('xy_cluster', XYClusterTransformer()),
    ('column_transform', ColumnTransformer([
        ('std', trans_stand, lista_stand),
        ('enc', trans_categoricas, lista_categoricas),
        ('tmp', trans_temporales, lista_temporales),
        ('con', trans_convertir, lista_convertir)
    ]))
])  

preprocesamiento.set_output(transform='pandas')


Pipeline(steps=[('inv', InvertirTransformer()),
                ('brand_cluster', BrandClusterTransformer()),
                ('xy_cluster', XYClusterTransformer()),
                ('column_transform',
                 ColumnTransformer(transformers=[('std',
                                                  Pipeline(steps=[('simple_imputer',
                                                                   SimpleImputer()),
                                                                  ('MinMaxScaler',
                                                                   MinMaxScaler())]),
                                                  ['num_deliver_per_week',
                                                   'transaction_count',
                                                   'recompra_promedio']),
                                                 ('enc',
                                                  Pipel...
                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                 sparse_output=False))]),
                                                  ['package', 'customer_type',
                                                   'segment', 'brand_cluster',
                                                   'XY_cluster']),
                                                 ('tmp',
                                                  Pipeline(steps=[('Date_Parser',
                                                                   DateParser())]),
                                                  ['year_week']),
                                                 ('con',
                                                  Pipeline(steps=[('Convertir',
                                                                   CategoricaSize()),
                                                                  ('OneHotEncoder_size',
                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                 sparse_output=False))]),
                                                  ['size'])]))])

In [40]:
df_transformado = preprocesamiento.fit_transform(df_completo)

c:\Users\naach\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning:

KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.



In [41]:
df_transformado.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9002580 entries, 6776133 to 9002579
Data columns (total 30 columns):
 #   Column                                     Dtype  
---  ------                                     -----  
 0   std__num_deliver_per_week                  float64
 1   std__transaction_count                     float64
 2   std__recompra_promedio                     float64
 3   enc__package_BOTELLA                       float64
 4   enc__package_KEG                           float64
 5   enc__package_LATA                          float64
 6   enc__customer_type_ABARROTES               float64
 7   enc__customer_type_CANAL FRIO              float64
 8   enc__customer_type_MAYORISTA               float64
 9   enc__customer_type_MINIMARKET              float64
 10  enc__customer_type_RESTAURANT              float64
 11  enc__customer_type_SUPERMERCADO            float64
 12  enc__customer_type_TIENDA DE CONVENIENCIA  float64
 13  enc__segment_HIGH                   

## 📌 Baseline [0.25 puntos]

<center>
<img src="https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExN3lzeGFqZmU3NzJrZHllNjRmaHVzczJpZ29rdHdlMzVpZnQwNXo1diZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/qAtZM2gvjWhPjmclZE/giphy.gif" width="300" height="200">

En esta sección se debe construir el modelo más sencillo posible que pueda resolver el problema planteado, conocido como **Modelo baseline**. Su propósito es servir como referencia para comparar el rendimiento de los modelos más avanzados desarrollados en etapas posteriores.  

Pasos requeridos:  
- Implemente, entrene y evalúe un modelo básico utilizando un pipeline.  
- Asegúrese de incluir en el pipeline las transformaciones del preprocesamiento realizadas previamente junto con un clasificador básico.  
- Evalúe el modelo y presente el informe de métricas utilizando **`classification_report`**.  

Documente claramente cómo se creó el modelo, las decisiones tomadas y los resultados obtenidos. Este modelo será la base comparativa en las secciones posteriores.

* Se comienzan separando los df en X e y, según cada caso (train, val y test).

In [50]:
# Se obtiene X e y en train, test y val
y_train = df_train['y']
X_train = df_train.drop(columns=['y'])

y_val = df_val['y']
X_val = df_val.drop(columns=['y'])

y_test = df_test['y']
X_test = df_test.drop(columns=['y'])

* Luego, en base a el conocimiento adquirido en laboratorios pasados y sabiendo que Dummy Classifier es un modelo de clasificación simple que genera predicciones en base a la estrategia definida, pero sin aprender de los datos, se elige como modelo baseline. En particular, la estrategia a seguir es 'stratified' ya que los datos se encuentran desbalanceados, por lo que esto respetará este comportamiento, observando así un modelo simple, el cual idealmente debe ser superado por modelos mucho más complejos en donde en el entrenamiento si aprenden de los datos.

In [43]:
from sklearn.dummy import DummyClassifier

# Se construye el baseline 
# Se juntan los pasos en un pipeline y se entrena
pipeline_dummy = Pipeline([('preprocesamiento', preprocesamiento),
                     ('dummy', DummyClassifier(strategy='stratified'))])

pipeline_dummy.fit(X_train, y_train)

c:\Users\naach\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning:

KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.



Pipeline(steps=[('preprocesamiento',
                 Pipeline(steps=[('inv', InvertirTransformer()),
                                 ('brand_cluster', BrandClusterTransformer()),
                                 ('xy_cluster', XYClusterTransformer()),
                                 ('column_transform',
                                  ColumnTransformer(transformers=[('std',
                                                                   Pipeline(steps=[('simple_imputer',
                                                                                    SimpleImputer()),
                                                                                   ('MinMaxScaler',
                                                                                    MinMaxScaler())]),
                                                                   ['num_deliver_per_week',
                                                                    'transaction_count...
                                                                                                  sparse_output=False))]),
                                                                   ['package',
                                                                    'customer_type',
                                                                    'segment',
                                                                    'brand_cluster',
                                                                    'XY_cluster']),
                                                                  ('tmp',
                                                                   Pipeline(steps=[('Date_Parser',
                                                                                    DateParser())]),
                                                                   ['year_week']),
                                                                  ('con',
                                                                   Pipeline(steps=[('Convertir',
                                                                                    CategoricaSize()),
                                                                                   ('OneHotEncoder_size',
                                                                                    OneHotEncoder(handle_unknown='ignore',
                                                                                                  sparse_output=False))]),
                                                                   ['size'])]))])),
                ('dummy', DummyClassifier(strategy='stratified'))])

* A continuación, se obtiene la predicción del modelo a partir de y_test, obteniendo un reporte del desempeño del modelo del baseline:

In [44]:
from sklearn.metrics import classification_report

# Se obtiene el clasification_report
y_pred = pipeline_dummy.predict(X_test)

print('Reporte de desempeño del modelo dummy: ')
print(classification_report(y_test, y_pred))

Reporte de desempeño del modelo dummy: 
              precision    recall  f1-score   support

           0       0.97      0.98      0.98   1749357
           1       0.03      0.02      0.02     51159

    accuracy                           0.95   1800516
   macro avg       0.50      0.50      0.50   1800516
weighted avg       0.94      0.95      0.95   1800516



* Observando los resultados obtenidos en las distintas métricas y el conocimiento en el desbalance en las clases, este comportamiento se ve reflejado en la clase mayoritaria con métricas bastante cercanas a 1 en todos los casos, pero si se observa la clase minoritaria, las métricas se acercan bastante a cero, por lo que es necesario encontrar modelos que puedan trabajar con este problema. 

## 📌 Elección de modelo [0.75 puntos]

En esta sección deben escoger un modelo que se adapte a las necesidades del negocio. Para esto, pruebe al menos 3 modelos y desarrolle los siguientes aspectos para cada uno:

- **Estructura y diferencias entre los modelos**: Explicar brevemente cada uno y sus hipérparámetros de mayor importancia.
- **Clasificadores recomendados**:
  - `LogisticRegression`
  - `KNeighborsClassifier`
  - `DecisionTreeClassifier`
  - `SVC`
  - `RandomForestClassifier`
  - `LightGBMClassifier` (del paquete `lightgbm`)
  - `XGBClassifier` (del paquete `xgboost`)
  - Otro (según lo que se estime adecuado)
  
- **Evaluación de resultados**: Se utilizará el **`classification_report`** para evaluar el rendimiento de cada modelo, destacando métricas clave como precisión, recall y F1-score. **Importante: No optimicen hiperparámetros, la idea es hacer una selección rápida del modelo.**

**Nota:** Pueden ocupar mas de 1 **instancia** de modelo para resolver el problema (e.g: (modelo_1, grupo_1), (modelo_2, grupo_2), ...).
  
A continuación, se deben responder las siguientes preguntas para evaluar el rendimiento de los modelos entrenados:

1. ¿Hay algún clasificador que supere al modelo baseline?  
2. ¿Cuál es el mejor clasificador entrenado y por qué?  
3. ¿Qué factores hacen que el mejor clasificador sea superior a los otros?  
4. En términos de `tiempo de entrenamiento`, ¿Qué modelo considera más adecuado para experimentar con grillas de optimización?

* En base a los modelos propuestos la elección de 3 de estos se decidió por las siguientes razones:
    * LogisticRegression: es un modelo que se suele utilizar en clasificación binaria y al trabajar con probabilidades, presenta mayor flexibilidad a la hora de clasificar según cada clase, pudiendo obtener predicciones más acertadas.
    * RandomForestClassifier: es un modelo bastante conocido para entrenar, el cual se conoce por tener un buen desempeño en general, debido a el entrenamiento con múltiples árboles, en vez de uno, lo que permite obtener resultados más precisos, pero teniendo en cuenta que este modelo también tiene un tiempo alto de entrenamiento.
    * XGBoost: teniendo tanto LGBM como XGB se elige el segundo, ya que se obtienen resultados similares en las predicciones, con una leve diferencia en el tiempo de entrenamiento, pero en comparación con los elegidos, este es mucho menor. Además, es un modelo que se encuentra altamente optimizado en tiempo y nivel de predicción, por lo que es un buen modelo a comparar con los otros 2.

* A continuación, se explican brevemente los modelos y sus hiperparámetros más importantes:
    * LogisticRegression:
        * Descripción: es un modelo que se utiliza en clasificación para predecir de manera binaria (la mayoría de las veces) la probabilidad de que una instancia sea parte de una clase o no, entregando un 1 si la probabilidad supera cierta probabilidad de pertenencia o un 0 en caso contrario.
        * Hiperparámetros: 
            - penalty: penalización que se aplicará al modelo
            - C: inverso de la regularización que se aplicará en el modelo para penalizar complejidad
            - solver: algoritmo que se utilizará para optimizar el problema
            - mas_iter: número máximo de iteraciones para que los solver convergan
    * RandomForestClassifier:
        * Descripción: es un modelo que utiliza un conjunto de árboles en diferentes subconjuntos para mejorar la precision en la clasificación, utilizando el promedio de los resultados obtenidos.
        * Hiperparámetros: 
            - n_estimator: número de árboles a utilizar en el modelo
            - max_depth: profundidad máxima a la que llegarán los árboles
            - min_samples_split: el número mínimo de muestras para que se divida un nodo interno
            - max_features: el número máximo de features a considerar a la hora de hacer cada división

    * XGBoostClassifier: 
        * Descripción: es un modelo de clasificación basado en gradient boosting que se encuentra altamente optimizado, donde cada nuevo árbol que se crea, corrige los errores que presentaron los anteriores
        * Hiperparámetros:
            - n_estimators: número de árboles a utilizar 
            - learning_rate: taza de aprendizaje que se aplicará en cada ajuste
            - max_depth: profundidad máxima a la que llegarán los árboles
            - subsample: cantidad de muestras a utilizar en cada árbol

* A continuación, se correrán los modelos explicados con anterioridad, utilizando los hiperparámetros predeterminados por el mismo modelo, para así poder observar el desempeño en su forma más 'básica':

In [45]:
# Se implementa LogisticRegression
from sklearn.linear_model import LogisticRegression

# Se juntan los pasos en un pipeline y se entrena
pipeline_LR = Pipeline([('preprocesamiento', preprocesamiento),
                     ('LogisticRegression', LogisticRegression())])

pipeline_LR.fit(X_train, y_train)

# Se obtiene el clasification_report
y_pred_LR = pipeline_LR.predict(X_test)

print('Reporte de desempeño del modelo Logistic Regression: ')
print(classification_report(y_test, y_pred_LR))

c:\Users\naach\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning:

KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.

c:\Users\naach\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning:

lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression



Reporte de desempeño del modelo Logistic Regression: 


c:\Users\naach\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1509: UndefinedMetricWarning:

Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.

c:\Users\naach\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1509: UndefinedMetricWarning:

Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.



              precision    recall  f1-score   support

           0       0.97      1.00      0.99   1749357
           1       0.00      0.00      0.00     51159

    accuracy                           0.97   1800516
   macro avg       0.49      0.50      0.49   1800516
weighted avg       0.94      0.97      0.96   1800516



c:\Users\naach\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1509: UndefinedMetricWarning:

Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.



* Según las métricas obtenidas para el modelo, se puede notar como la clase mayoritaria se hace presente en las predicciones con valores bastante cercanos, e incluso iguales a 1. Por otro lado, en la clase minoritaria los valores son simplemente 0, por lo que da una alerta del modelo, ya que aunque mejora levemente en las métricas asociadas a la clase 0, en comparación con el baseline, cae a 0 en las predicciones de la clase 1.

In [46]:
# Se implementa RandomForest
from sklearn.ensemble import RandomForestClassifier

# Se juntan los pasos en un pipeline y se entrena
pipeline_RFC= Pipeline([('preprocesamiento', preprocesamiento),
                     ('random forest classifier', RandomForestClassifier())])

pipeline_RFC.fit(X_train, y_train)

# Se obtiene el clasification_report
y_pred_RFC= pipeline_RFC.predict(X_test)

print('Reporte de desempeño del modelo Random Forest Classifier: ')
print(classification_report(y_test, y_pred_RFC))

c:\Users\naach\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning:

KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.



Reporte de desempeño del modelo Random Forest Classifier: 
              precision    recall  f1-score   support

           0       0.98      0.99      0.99   1749357
           1       0.50      0.21      0.30     51159

    accuracy                           0.97   1800516
   macro avg       0.74      0.60      0.64   1800516
weighted avg       0.96      0.97      0.97   1800516



* Según las métricas obtenidas para el modelo, se puede notar que ahora hay una mejora considerable en las predicciones de la clase 1, aumentando a incluso valores de precision iguales a 0.5, recall 0.21 y f1-score 0.3. Además, en comparación al baseline, las métricas asociadas a la clase 0, se mantienen iguales, por lo que de manera general, se puede notar una mejora en el modelo random forest en comparación al baseline.

In [54]:
# Se implementa XGBoost classifier
from xgboost import XGBClassifier
# Se juntan los pasos en un pipeline y se entrena
pipeline_XGBC= Pipeline([('preprocesamiento', preprocesamiento),
                     ('xgboost classifier', XGBClassifier())])

pipeline_XGBC.fit(X_train, y_train)

# Se obtiene el clasification_report
y_pred_XGBC= pipeline_XGBC.predict(X_test)

print('Reporte de desempeño del modelo XGBoost Classifier: ')
print(classification_report(y_test, y_pred_XGBC))

c:\Users\naach\anaconda3\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning:

KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.



Reporte de desempeño del modelo XGBoost Classifier: 
              precision    recall  f1-score   support

           0       0.98      1.00      0.99   1749357
           1       0.65      0.15      0.24     51159

    accuracy                           0.97   1800516
   macro avg       0.81      0.57      0.61   1800516
weighted avg       0.97      0.97      0.97   1800516



* Según las métricas obtenidas para el modelo, también se pueden observar mejorar con respecto al baseline y leves diferencias si se compara con Random Forest, ya que mejora en precision, pero si se observa el f1-score, en el caso de xgboost, este valor cae.

1. ¿Hay algún clasificador que supere al modelo baseline?  
    - Al probar 3 de los modelos propuestos, 2 de ellos superan en la predicción al modelo baseline, los cuales son Random Forest y XGBoost. 
    - Se observa principalmente la mejora en la predicción de la clase 1, aumentando considerablemente los valores cercanos a cero obtenidos en las métricas precision, recall y f1-socre.

2. ¿Cuál es el mejor clasificador entrenado y por qué?  
    - Si se observan los clasificadores que superaron al baseline, se puede notar que obtienen un desempeño bastante similar, pero como es necesario observar todas las métricas para poder entender el desempeño de estos, es que se pone atención en f1-score, donde en el caso de Random Forest se obtiene un mejor equilibrio entre las métricas precision y recall. Por lo anterior, se determina que el mejor modelo es Random Forest, pero es importante tener en cuenta que esta leve mejora viene de la mano con un tiempo de entrenamiento mucho mayor.

3. ¿Qué factores hacen que el mejor clasificador sea superior a los otros?  
    - Tal como se mencionó antes, se observa la métrica f-score para poder observar el comportamiento de ambas métricas (precision y recall) y así tomar en consideración tanto el comportamiento con los falsos negativos como con los positivos para decidirse por el mejor

4. En términos de `tiempo de entrenamiento`, ¿Qué modelo considera más adecuado para experimentar con grillas de optimización?
    - En relación al tiempo de entrenamiento y considerando que en los modelos que superan al baseline no hay una gran diferencia en sus métricas, se considera más adecuado XGBoost para experimentar con grillas, ya que al estar más optimizado, el tiempo que demora en el entrenamiento es mucho menor.

## 📌 Optimización de Hiperparámetros [1.0 puntos]

<center>
<img src="https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExcXJkNzdhYjlneHplaGpsbnVkdzh5dnY3Y2VyaTIzamszdGR1czJ2diZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/2rqEdFfkMzXmo/giphy.gif" width="300" height="200">

A partir de su análisis anterior, se debe proceder a optimizar el rendimiento del modelo seleccionado mediante la optimización de sus hiperparámetros. Para ello, se espera que implementen `Optuna` para optimizar no solo los hiperparámetros del modelo, sino también los de los preprocesadores utilizados (por ejemplo, OneHot Encoding, Scalers, etc.).

Al desarrollar este proceso, deberán responder las siguientes preguntas clave como mínimo:

- ¿Qué métrica decidieron optimizar y por qué?

- ¿Qué hiperparámetro tuvo un mayor impacto en el rendimiento de su modelo?

- ¿Cuánto mejoró el rendimiento del modelo después de la optimización de hiperparámetros?

In [ ]:
import optuna
from optuna.samplers import TPESampler
optuna.logging.set_verbosity(optuna.logging.WARNING)
from sklearn.metrics import mean_absolute_error


def objective(trial):
    # Hiperparámetros que se pide optimizar con sus rangos
    learning_rate = trial.suggest_float("learning_rate", 0.001, 0.1)
    n_estimators = trial.suggest_int("n_estimators", 50, 2000)
    max_depth = trial.suggest_int("max_depth", 3, 20)
    max_leaves = trial.suggest_int("max_leaves", 0, 100)
    min_child_weight = trial.suggest_int("min_child_weight", 1, 5)
    reg_alpha = trial.suggest_float("reg_alpha", 0.0, 1.0)
    reg_lambda = trial.suggest_float("reg_lambda", 0.0, 1.0)

    # Simple Imputer
    strategy = trial.suggest_categorical("strategy", ["mean", "median", "most_frequent", "constant"])

    # MinMax Scaler
    feature_range = trial.suggest_categorical("feature_range", [(0, 1), (-1, 1), (0, 10)])

    # One Hot Encoder
    min_freq = trial.suggest_float("min_frequency", 0.0, 1.0) 

    
    # Se asignan a los preprocesadores, los hiperparametros a optimizar
    trans_stand = Pipeline([
        ('simple_imputer', SimpleImputer(strategy=strategy)), 
        ('MinMaxScaler', MinMaxScaler(feature_range=feature_range))
    ])

    trans_categoricas = Pipeline([
        ('OneHotEncoder', OneHotEncoder(sparse_output=False, handle_unknown='ignore', min_frequency=min_freq))
    ])

    trans_temporales = Pipeline([
        ('Date_Parser', DateParser())
    ])

    trans_invertir = Pipeline([
        ('Invertir', InvertirTransformer())
    ])

    trans_convertir = Pipeline([
        ('Convertir',CategoricaSize()),
        ('OneHotEncoder_size', OneHotEncoder(sparse_output=False, handle_unknown='ignore', min_frequency=min_freq))
    ])

    preprocesamiento = Pipeline([
        ('inv', InvertirTransformer()),
        ('brand_cluster', BrandClusterTransformer(n_clusters=3)),
        ('xy_cluster', XYClusterTransformer()),
        ('column_transform', ColumnTransformer([
            ('std', trans_stand, lista_stand),
            ('enc', trans_categoricas, lista_categoricas),
            ('tmp', trans_temporales, lista_temporales),
            ('con', trans_convertir, lista_convertir)
        ]))
    ])  

    preprocesamiento.set_output(transform='pandas')

    # Entrenamos con lo optimizado
    model = XGBClassifier(
        learning_rate=learning_rate,
        n_estimators=n_estimators,
        max_depth=max_depth,
        max_leaves=max_leaves,
        min_child_weight=min_child_weight,
        reg_alpha=reg_alpha,
        reg_lambda=reg_lambda,
        random_state=42,
        verbosity=0
    )

    # Pipeline con el modelo y transformación
    pipeline = Pipeline([
        ('preprocesamiento', preprocesamiento),
        ('xgboost classifier', model)
    ])

    pipeline.fit(X_train, y_train)


    y_pred = pipeline.predict(X_test)
    mae = mean_absolute_error(y_test, y_pred)

    # Guardamos
    trial.set_user_attr("best_pipeline", pipeline)

    return mae

In [ ]:
#Finamos una semilla para reproducibilidad
sampler = TPESampler(seed=42)
study = optuna.create_study(direction="minimize", sampler=sampler)
study.optimize(objective, timeout=300)  # por 5 minutos

In [ ]:
print("Número de trials:", len(study.trials))
print("Mejor MAE:", study.best_value)
print("Mejores hiperparámetros:")
for key, value in study.best_params.items():
    print(f"  {key}: {value}")

In [ ]:
# Guardamos el modelo
model_xgb_optuna = study.best_trial.user_attrs["best_pipeline"]

In [ ]:
# Se obtiene el clasification_report
y_pred_XGBC_optuna= model_xgb_optuna.predict(X_test)

print('Reporte de desempeño del modelo XGBoost Classifier: ')
print(classification_report(y_test, y_pred_XGBC_optuna))

## 📌 Interpretabilidad [1.0 puntos]

En esta sección, deben explicar el funcionamiento de su modelo utilizando las técnicas de interpretabilidad vistas en clase, como `SHAP`. Se espera que sean capaces de descomponer las predicciones y evaluar la importancia de los atributos y las interacciones entre ellos, con el fin de obtener una comprensión más profunda de cómo el modelo toma decisiones. 

Al desarrollar esta parte, deberán responder las siguientes preguntas clave como mínimo:

- ¿Podría explicar el funcionamiento de su modelo para una predicción en particular? Si es así, proporcione al menos tres ejemplos específicos, describiendo cómo el modelo llegó a sus decisiones y qué factores fueron más relevantes en cada caso.

- ¿Qué atributo tiene una mayor importancia en la salida de su modelo? Analice si esto tiene sentido con el problema planteado y justifique la relevancia de dicho atributo en el contexto de las predicciones que se realizan.

- ¿Existe alguna interacción entre atributos que sea relevante para el modelo? Investigue si la combinación de ciertos atributos tiene un impacto significativo en las predicciones y explíquela en **detalle**.

- ¿Podría existir sesgo hacia algún atributo en particular? Reflexione sobre la posibilidad de que el modelo esté favoreciendo ciertos atributos. Si es así, ¿cuál podría ser la causa y qué impacto podría tener esto en la predicción?

* Se comenzaran observando las importancias de las features de manera global:

In [ ]:
import numpy as np
from sklearn.inspection import permutation_importance

# Se obtiene el orden de las features
order = model_xgb_optuna.named_steps['xgboost classifier'].get_booster().feature_names

# Se implementa un metodo de permutacion para observar la importancia de features
result = permutation_importance(model_xgb_optuna, X_test, y_test, n_repeats=1, random_state=42)

# Se visualiza en un grafico
perm_sorted_idx = result.importances_mean.argsort()

new_order = []
for i in range(perm_sorted_idx.shape[0]-1, -1, -1):
    new_order = np.append(new_order, order[perm_sorted_idx[i]])

plt.boxplot(
    result.importances[perm_sorted_idx].T,
    vert=False,
    labels = new_order
)
plt.title("Importancia de las Features por Incremento del Error")
plt.xlabel('Error')
plt.show()

* Si se observa el gráfico de importancia ...

* Ahora, se intentará explicar el funcionamiento del modelo para ciertas predicciones:

In [ ]:
# !pip install shap

In [ ]:
# Se define el explainer del mdelo
import shap 

# Se transforma X_test para utilizar en explainer, ya que explainer solo acepta el modelo xgboost
X_train_transformado = preprocesamiento.fit_transform(X_train)
X_test_transformado = preprocesamiento.transform(X_test)
explainer = shap.TreeExplainer(pipeline_XGBC.named_steps['xgboost classifier'])
shap_values = explainer(X_test_transformado)

# transformar logits a probabilidad
np.exp(shap_values.base_values) / (1 + np.exp(shap_values.base_values))

In [ ]:
# Primer ejemplo
idx_muestra = 1

shap.initjs()
shap.force_plot(explainer.expected_value, 
                shap_values.values[idx_muestra,:], 
                X_test_transformado.iloc[idx_muestra,:],
                link="logit")

In [ ]:
# Segundo ejemplo
idx_muestra = 2

shap.initjs()
shap.force_plot(explainer.expected_value, 
                shap_values.values[idx_muestra,:], 
                X_test_transformado.iloc[idx_muestra,:],
                link="logit")

In [ ]:
# Tercer ejemplo
idx_muestra = 3

shap.initjs()
shap.force_plot(explainer.expected_value, 
                shap_values.values[idx_muestra,:], 
                X_test_transformado.iloc[idx_muestra,:],
                link="logit")

## 📌 Resultados y Conclusiones [1.0 puntos]

Para finalizar, se deben explicar los desarrollos y resultados obtenidos a lo largo de todo el proceso, desde la selección de las variables hasta la optimización de hiperparámetros e interpretación. Se espera una reflexión crítica sobre el desempeño de los modelos entrenados y una comparación entre los diferentes enfoques. Además, deberán abordar los siguientes puntos clave:

- **Análisis de métricas**: Comenten sobre las métricas obtenidas en cada etapa del modelo, destacando las más relevantes como precisión, recall, F1-score, etc. ¿Cuáles fueron los modelos más efectivos? ¿Qué diferencias notables encontró entre ellos?

- **Impacto de las decisiones tomadas**: Reflexionen sobre cómo las decisiones relacionadas con el preprocesamiento, selección de atributos y optimización de hiperparámetros influyeron en los resultados finales. ¿Hubo alguna decisión que haya tenido un impacto notable en el rendimiento?

- **Lecciones aprendidas**: Concluyan sobre las lecciones más importantes que aprendieron durante el proceso y cómo estas pueden influir en futuras iteraciones del modelo. ¿Qué se podría mejorar si se repitiera el proceso? Si tuvieran más recursos y tiempo, ¿qué otras técnicas/herramientas habrían utilizado?

> [Escriba aquí sus resultados]

Mucho éxito!

<center>
<img src="https://media3.giphy.com/media/v1.Y2lkPTc5MGI3NjExaHpvOTY5Z3hpdHI3aDBpdGRueXRqamZncXp2emFrbjJ5M2s5eTR1dSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/1PMVNNKVIL8Ig/giphy.gif" width="300" height="200">
